donkeycar/parts/gps.py

```diff
@@ -0,0 +1,641 @@
+        #
+        if debug:
+            try:
+                msg = pynmea2.parse(line)
+                logger.info(f"nmea.longitude({msg.longitude}, nmea.latitude({msg.latitude})")
+            except pynmea2.ParseError as e:
+                logger.info('Ignoring NMEA parse error: {}'.format(e))
+
+        # Reading the GPS fix data is an alternative approach that also works
+        if nmea_parts[2] == 'V':
+            # V = Warning, most likely, there are no satellites in view...
+            logger.warning("GPS receiver warning; position not valid")
+        else:
+            #
+            # Convert the textual nmea position into degrees
+            #
+            longitude = nmea_to_degrees(nmea_parts[5], nmea_parts[6])
+            latitude = nmea_to_degrees(nmea_parts[3], nmea_parts[4])
+            # print(f"Your position: lon = ({longitude}), lat = ({latitude})")
+            
+            #
+            # convert position in degrees to local meters
+            #
+            utm_position = utm.from_latlon(latitude, longitude)
+            # print(f"Your utm position: lon - ({utm_position[1]:.6f}), lat = ({utm_position[0]:.6f})")
+            
+            # return (longitude, latitude) as float degrees
+            return(utm_position[1], utm_position[0])
+    else:
+        # Non-position message OR invalid string
+        # print(f"Ignoring line {line}")
+        pass
+    return None
+
+
+def parse_nmea_checksum(nmea_line):
+    """
+    Given the complete nmea line (including starting '$' and ending checksum '*##')
+    calculate the checksum from the body of the line.
+    NOTE: this does not check for structural correctness, so you
+          should check that '$' and '*##' checksum are present before
+          calling this function.
+    """
+    return int(nmea_line[-2:], 16) # checksum hex digits as int
+    
+    
+def calculate_nmea_checksum(nmea_line):
+    """
+    Given the complete nmea line (including starting '$' and ending checksum '*##')
+    calculate the checksum from the body of the line.
+    NOTE: this does not check for structural correctness, so you
+          should check that '$' and '*##' checksum are present
+          and that the checksum matches before calling this function.
+    """
+    # 
+    # xor all characters in the message to get a one byte checksum.
+    # don't include starting '$' or trailing checksum '*##'
+    #
+    return reduce(operator.xor, map(ord, nmea_line[1:-3]), 0)
+
+
+def nmea_to_degrees(gps_str, direction):
+    """
+    Convert a gps coordinate string formatted as:
+    DDMM.MMMMM, where DD denotes the degrees and MM.MMMMM denotes the minutes
+    to a float in degrees.
+    """
+    if not gps_str or gps_str == "0":
+        return 0
+        
+    #
+    # pull out the degrees and minutes
+    # and then combine the minutes
+    #
+    parts = gps_str.split(".")
+    degrees_str = parts[0][:-2]
+    minutes_str = parts[0][-2:]
```

Comment:
No that is not quite right.  the format is DDMM.MMM.  In those likes I am separating the the whole degrees from the whole minutes.  I guess I should right a unit test for this, but I did check it agains the pynmea library's output and it matches.

Reply:
I did it this way because there are a variable number of degree digits, but always 2 whole minute digits.  `parts[0][:-2]` results in zero to 3 digits.  I need to do the split to get that to work.

Reply:
I added comments to make that clear.